### PR TITLE
Replace libLLVM symlink with linker script

### LINKED ISSUE #121889: Miri (possibly everything that depends on rustc private crates) no longer builds: cannot find -lLLVM-18-rust-1.78.0-nightly
Trying to build Miri against latest rustc artifacts fails:
```
error: linking with `cc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/r/.opam/coq-8.17/bin:/home/r/bin:/home/r/.cargo/bin:/home/r/.local/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games" VSLANG="1033" "cc" "-m64" "/tmp/rustcKdET34/symbols.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.11xmci8slcb5avxi.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.12qdf7a036c4vr02.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.15b6acljysgkq0tu.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.16aneeg9x35abgfk.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.19n04kvindlihsq4.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1anbmrreqp5yl3t7.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1cvirxj3fkpsytp3.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1f2f3tzhb4qd756u.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1fx8rr91ki8r0rlb.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1gcfbjf0sqqsikft.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1hsv5pb43rvh4f79.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1jv2r4bh2poygsjx.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1jylhxsdskbbm99x.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1k5lzruxwatzbj2s.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1mgar9vu7xj3r5mb.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1mv2ygq8yvvq0jq6.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1pkdaqyh21r0839z.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1rioyg5mmyf8ssnd.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1t27p4z7bidzdjqd.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1tkxukot08dgpwbt.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1tz9gmedjn9p737z.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1v72fc0np44swbku.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1w1ju2tfjtpulam9.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1wd50i4kd74vdv4c.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.1xwt5aalooldi3ex.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.20dyrydgf428k5q5.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.211hbbml25qq68ez.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.23gucqwjf7j5l69x.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.24ib78nyvdpsehn2.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.29cxicvofkw0jq4z.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.29x9mfivg0jv356x.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2a3ksyit21txexs0.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2jjqzvqhtzu921ja.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2mmwcx28xlg7n3y.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2nhvryvgosli4t2b.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2rzxpo4wjlg688le.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2u52aaxbbnmzirqq.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2w2dng6cmtx51n0t.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2y1135tvki6xx0us.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.2zbm1kexdud4ock.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.31lql8y4cwqq1rsc.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.33ucu751hxg4cyp1.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.34zsrisfrvqit1qq.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.35311tvn2k2eznof.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.35glmbccb6dlrm6t.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.37242ih5zl9hnflb.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3frlgq4ppsqjfetr.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3kjz6fj8y8a3b2lq.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3m5ddhkq79a4w9zn.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3neuj8d1sh5n5y70.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3srbft7cf0q3knwl.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3y1du7vm261s6dqc.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.3za3p4ouch3zlwft.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.40aepun10jbxed3l.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.40s7n70at8yvyzky.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.41vkwl7szymn0ydb.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.42cxnfc26ifudjg4.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.430qo95zjwl988gu.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.45ghfp84t7v6v1zn.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.46n99y900bjdc2c9.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.46vcmvebey20dmo2.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.47unlhthlw5w3ur4.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.47zeayp8oss7h8nj.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4bnhetudb5on11hn.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4fjwgz4uzeevtuym.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4g2yyfwyz2o8orwf.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4gac27tdcsxhmsg8.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4gnz9hfs1leuyg3v.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4h8f28uyw22iopxn.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4iisxetqadowrrjc.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4jn1wt1wyf6k5uqv.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4opzxpspbu41ps9j.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4qb2rbmcieiz0lkv.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4qw6owwety6j8tti.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.4ws90etbifrlzvyp.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.54j4girgkeq887ai.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.54ozlt8pgqdcir2n.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.55snzhhyjc1y3xuw.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.56fyqr7ezx9hnynk.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.5gm2kf7lr0z2mz4z.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.63wn2sw0lsflw4q.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.6bvsrczg4mob2iq.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.bvzhfeqrd1miunh.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.gfgkbg63bdgtrhu.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.hcsfaat1yr86gdb.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.hwhafoy2x1yrwu9.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.iece6mxl3ynkwfm.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.m1ycnnzho3hu308.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.nazzxdvx4tni2st.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.qq8qh95ksor3yte.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.vtnwx071pemu57t.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.ypwq9ji9yo78kzz.rcgu.o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e.ysmoy9rm9wap8g2.rcgu.o" "-Wl,--as-needed" "-L" "/home/r/src/rust/miri.2/target/debug/deps" "-L" "/home/r/src/rust/miri.2/target/debug/build/jemalloc-sys-e486b8af66a531e2/out/build/lib" "-L" "/home/r/src/rust/miri.2/target/debug/build/libffi-sys-e1e746640b368ab4/out/libffi-root/lib" "-L" "/home/r/src/rust/miri.2/target/debug/build/libffi-sys-e1e746640b368ab4/out/libffi-root/lib64" "-L" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/r/src/rust/miri.2/target/debug/deps/libaes-d27d768da7a7b0b1.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libcpufeatures-761c6e9bf04d6606.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libcipher-c90706229fcff33e.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libinout-de362980c6a24efa.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libcrypto_common-d0f50804162871cb.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libgeneric_array-3f2ffc280ce1dbbf.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libtypenum-f28a8382956c41d8.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblibloading-306f9ee4a9858b9d.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libmeasureme-2d58b4bc5bbf7ebe.rlib" "/home/r/src/rust/miri.2/target/debug/deps/librustc_hash-bbbf74eb674bc8eb.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libparking_lot-2a8e11285ae5880f.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libparking_lot_core-82788d9c72c2ac23.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblock_api-3f7a67bc67bad859.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libscopeguard-0e29fa3dc136589b.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libperf_event_open_sys-796a3bcc04573a4f.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libmemmap2-42ecdbe0a18eebc7.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblog-bffec5f23a73e889.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libctrlc-2c97869c38cdbe01.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libnix-5222d02097068a7d.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libbitflags-307cc57d485557b2.rlib" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib/libtest-0cd255ee414ee1b7.rlib" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-d94d981cfad2e302.rlib" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunicode_width-bc5c85135ebc2d28.rlib" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_std-a783581261810918.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblibffi-1d4f6b13e9fd3e81.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblibffi_sys-99c31e51ef500979.rlib" "/home/r/src/rust/miri.2/target/debug/deps/librand-88ef0f10864fca5a.rlib" "/home/r/src/rust/miri.2/target/debug/deps/librand_chacha-6b0a1e7cb7bcc201.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libppv_lite86-fe09708be6de18e2.rlib" "/home/r/src/rust/miri.2/target/debug/deps/librand_core-d7a6061eaf088212.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libgetrandom-d5c3da24fb92135e.rlib" "/home/r/src/rust/miri.2/target/debug/deps/liblibc-433eaf9623d4a3c8.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libcfg_if-33350d16ae301f5d.rlib" "/home/r/src/rust/miri.2/target/debug/deps/libsmallvec-d31fd5d95e65d547.rlib" "-L" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bdynamic" "-lrustc_driver-c39020f115acfe15" "-L" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-lstd-41777efa78699460" "-Wl,-Bstatic" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-b7903030bc1640bf.rlib" "-Wl,-Bdynamic" "-ldl" "-lLLVM-18-rust-1.78.0-nightly" "-ldl" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/home/r/src/rust/miri.2/target/debug/deps/miri-c318c1261dd2c19e" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-Wl,-rpath,/home/r/.rustup/toolchains/miri/lib/rustlib/x86_64-unknown-linux-gnu/lib"
  = note: /usr/bin/ld: cannot find -lLLVM-18-rust-1.78.0-nightly: No such file or directory
          collect2: error: ld returned 1 exit status
```

Regression range: https://github.com/rust-lang/rust/compare/6f435eb0e...eaee1e945

Likely cause: https://github.com/rust-lang/rust/pull/121395
Cc @nikic

### PR DESCRIPTION
**Pull request by nikic**
It turns out that the libLLVM-N.so -> libLLVM.so.N.1 symlink is also needed when projects like miri link against librustc_driver.so. As such, we have to distribute it in real rustup components like rustc-dev, rather than only for download-ci-llvm.

To avoid actually distributing symlinks (which are not supported or not fully supported by the rustup infrastructure) replace it with a linker script that does the same thing instead.

Fixes https://github.com/rust-lang/rust/issues/121889.

r? @cuviper

### CODE REVIEW

File: src/bootstrap/src/core/build_steps/dist.rs [hunk 1]
```diff
@@ -2035,11 +2030,15 @@ fn install_llvm_file(
...
-        if install_symlink {
-            // If requested, also install the symlink. This is used by download-ci-llvm.
-            let full_dest = destination.join(source.file_name().unwrap());
-            builder.copy(&source, &full_dest);
-        }
+
+        // Replace the symlink itself with an equivalent linker script. This is used by
+        // download-ci-llvm and projects linking against librustc_driver.so. We don't use a
+        // symlink, as these are not allowed inside rustup components.
+        let link = t!(fs::read_link(source));
```
Comment (bjorn3) [inline src/bootstrap/src/core/build_steps/dist.rs:2037]: This will be a relative path, right?

Comment (nikic) [reply src/bootstrap/src/core/build_steps/dist.rs:2037]: Yes, this will produce something like `INPUT(libLLVM.so.18.1-rust-dev)`.

**Comment by nikic** [comment]
Breaks download-ci-llvm, because that doesn't have the symlink now either, so the new logic doesn't apply to it anymore.

I'll restore the symlink for the download-ci-llvm case.

**Review by nbdd0121 — APPROVED** [review]
Code looks good, although it's a bit sad that w have separate code path for download-ci-llvm and everything else.

**Comment by nikic** [comment]
I've confirmed that download-ci-llvm works again now, and also confirmed that miri can be built against this toolchain (I had to edit miri-script to disable rustfmt and clippy, which are not part of the try toolchain).

**Comment by nikic** [comment]
> > Breaks download-ci-llvm, because that doesn't have the symlink now either, so the new logic doesn't apply to it anymore.
> 
> Why does the linker script not work for download-ci-llvm?

It does work for the purposes of linking, but it breaks the code being modified here that handles libLLVM installation. The logic is based around having the symlink. If we don't ship the symlink for download-ci-llvm, then the logic would additionally also have to support reading the linker script to determine which shared object it needs to install in the libdir.

**Comment by RalfJung** [comment]
Oh, this function later works on its own output as input? Wow that's recursive.^^

Might be worth having a comment in the code explaining that.

**Comment by bjorn3** [comment]
On Windows 10 it still requires developer mode to be enabled, right?

**Comment by mati865** [comment]
Both Windows 10 and 11 require either developer mode or modification of group policies.

**Comment by RalfJung** [comment]
@cuviper this is blocking all tools linking against rustc_driver, would be nice to get it landed soon. :)

**Comment by Mark-Simulacrum** [comment]
@bors r+

The risk here seems fairly low (I expect to hit some problems but that's almost always true with changes like this). In the meantime, it should unblock miri etc.

**Pull request merged**